amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java

```diff
@@ -320,27 +332,32 @@ private void errorOccurred(RecordFlow triggeringFlow, Throwable t) {
     }
 
     // This method is not thread safe. This needs to be executed after acquiring lock on this.lockObject
-    private void clearRecordsDeliveryQueue(boolean isSubscriptionCompleting) {
-        // If the queue is empty, we can immediately return without any further action.
-        if (recordsDeliveryQueue.isEmpty()) {
-            return;
+    private void resetRecordsDeliveryStateOnSubscriptionInit() {
+        // Clear any lingering records in the queue.
+        if (!recordsDeliveryQueue.isEmpty()) {
+            recordsDeliveryQueue.clear();
         }
+        // Set pendingActiveSubscriptionShutdown to default value.
+        pendingActiveSubscriptionShutdown = false;
+    }
 
+    // This method is not thread safe. This needs to be executed after acquiring lock on this.lockObject
+    private void resetRecordsDeliveryStateOnSubscriptionShutdown() {
         // Wait for final event notification during the end of the subscription.
-        if(isSubscriptionCompleting) {
+        if (!recordsDeliveryQueue.isEmpty()) {
             // This will prevent further events from getting scheduled, during the wait period.
-            isAwaitingFinalAckForCurrentSubscription = true;
+            pendingActiveSubscriptionShutdown = true;
             try {
                 // Wait for the configured time to get a notification for already delivered event, if any.
                 lockObject.wait(TIME_TO_WAIT_FOR_FINAL_ACK_MILLIS);
             } catch (InterruptedException e) {
                 Thread.currentThread().interrupt();
             }
+            // Clear the queue to remove any remaining entries from the queue.
+            recordsDeliveryQueue.clear();
```

Comment:
Why cant we clear the queue before waiting ?

Reply:
If we clear before waiting, the ack received will be marked as stale and records will be duplicated.